src/apps/racks/services.py

```diff
@@ -0,0 +1,187 @@
+from decimal import Decimal
+from typing import Union
+
+from pydantic import BaseModel
+from sqlalchemy import delete, select, update
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from src.apps.racks.models import Rack
+from src.apps.racks.schemas import (
+    RackBaseOutputSchema,
+    RackInputSchema,
+    RackOutputSchema,
+    RackUpdateSchema,
+)
+from src.apps.sections.models import Section
+from src.apps.sections.services import manage_section_state
+from src.apps.warehouse.models import Warehouse
+from src.apps.warehouse.services import get_all_warehouses, manage_warehouse_state
+from src.core.exceptions import (
+    AlreadyExists,
+    DoesNotExist,
+    IsOccupied,
+    NotEnoughSectionResourcesException,
+    NotEnoughWarehouseResourcesException,
+    RackIsNotEmptyException,
+    ServiceException,
+    TooLittleRackLevelsAmountException,
+    TooLittleRacksAmountException,
+    TooLittleWeightAmountException,
+    WarehouseAlreadyExistsException,
+    WarehouseDoesNotExistException,
+)
+from src.core.pagination.models import PageParams
+from src.core.pagination.schemas import PagedResponseSchema
+from src.core.pagination.services import paginate
+from src.core.utils.orm import if_exists
+
+
+async def create_rack(
+    session: AsyncSession, rack_input: RackInputSchema
+) -> RackOutputSchema:
+    rack_data = rack_input.dict()
+    section_id = rack_data.get("section_id")
+    if not (section_object := await if_exists(Section, "id", section_id, session)):
+        raise DoesNotExist(Section.__name__, "id", section_id)
+
+    if not section_object.available_racks:
+        raise NotEnoughSectionResourcesException(resource="racks")
+
+    if (
+        rack_max_weight := rack_data.get("max_weight")
+    ) > section_object.available_weight:
+        raise NotEnoughSectionResourcesException(resource="available section weight")
```

Comment:
there will be checking of section_object.weight_to_reserve as creating empty rack is related to the amount of weight to reserve, not the available_weight_value (this applies to max_section_weight - occupied_weight (by stocks) )